### PR TITLE
assesment status

### PR DESCRIPTION
**Pull request by Devyani606**
ples chk

### CODE REVIEW

**Review by Devyani606 — COMMENTED** [review]
actually i m not getting the body response

File: config/application.yml [hunk 1]
```diff
@@ -9,3 +9,6 @@ AWS_ACCESS_KEY_ID: 'SDJFKSADSAD'
 AWS_SECRET_ACCESS_KEY: 'AaC+bADJJRyKYf7uV893WIUSADJH'
 AWS_REGION: 'ap-south-1'
 S3_BUCKET_URL: 'https://oncot.s3.ap-south-1.amazonaws.com'
+PARAMS_AI_URL: 'http://localhost:3001/api/v1/webhook'
```
Comment (somnath77) [inline config/application.yml:12]: I think `PARAMS_AI_URL` is not necessary

File: spec/controllers/api/v1/candidates_controller_spec.rb [hunk 2]
```diff
@@ -41,6 +45,17 @@
         expect(response).to have_http_status(:not_found)
       end
     end
+
+    context 'valid params with assessmnet worker' do
```
Comment (somnath77) [inline spec/controllers/api/v1/candidates_controller_spec.rb:49]: assessment spelling mistake

File: spec/controllers/api/v1/candidates_controller_spec.rb [hunk 3]
```diff
@@ -41,6 +45,17 @@
...
+      it 'get succesfully data for assesment' do
+        params = {
+          id: 10,
+          token: @drives_candidate2.token
+        }
```
Comment (somnath77) [inline spec/controllers/api/v1/candidates_controller_spec.rb:54]: lines 51 to 54 can be in single line

File: spec/controllers/api/v1/candidates_controller_spec.rb (hunk 2 above)
Comment (Devyani606) [reply spec/controllers/api/v1/candidates_controller_spec.rb:49]: done

File: spec/controllers/api/v1/candidates_controller_spec.rb (hunk 3 above)
Comment (Devyani606) [reply spec/controllers/api/v1/candidates_controller_spec.rb:54]: done

File: config/application.yml (hunk 1 above)
Comment (Devyani606) [reply config/application.yml]: done

**Review by somnath77 — APPROVED** [review]

File: app/controllers/api/v1/results_controller.rb [hunk 4]
```diff
@@ -28,22 +28,35 @@ def find_drive_candidates
...
-                    drives_candidate.score] + test_cases + [submission.answer]
+                    drives_candidate.score || 0] + test_cases + [submission.answer]
           end
+        else
+          drive_candidate = DrivesCandidate.where(id: params[:drives_candidate_id]).first
+          candidate = drive_candidate.candidate
+          submission = drive_candidate.submissions.first
+          test_cases = test_case_results(submission.id)
+          csv << [candidate.first_name, candidate.last_name, candidate.email,
+                  drive_candidate.score || 0] + test_cases + [submission.answer]
```
Comment (anilmaurya) [inline app/controllers/api/v1/results_controller.rb:59]: This is duplicate code which can be moved to another method and use at both places

File: app/workers/assessment_webhook_worker.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    drive_candidate = DrivesCandidate.where(id: drive_candidate_id).first
+    drive = drive_candidate.drive
+    return unless drive_candidate.present?
+
+    params = { key: ENV['PARAMS_AI_KEY'], assessment_schedule_id: drive_candidate.uuid, score: drive_candidate.score,
+               status: 'completed', report: "#{ENV['MAIN_URL']}/api/v1/drives/#{drive.id}/results/csv_result?drives_candidate_id=#{drive_candidate_id}" }
+    status = ParamAiApi.new(params).post
+    return unless status != 200
+
+    AssessmentWebhookWorker.perform_in(10.minutes, drive_candidate_id)
```
Comment (anilmaurya) [inline app/workers/assessment_webhook_worker.rb:16]: I think we can utilise sidekiq retry mechanism, let's raise exception here which sidekiq will catch and retry for 25 times with exponential backoff time.

File: app/controllers/api/v1/results_controller.rb (hunk 4 above)
Comment (Devyani606) [reply app/controllers/api/v1/results_controller.rb]: done

File: app/workers/assessment_webhook_worker.rb (hunk 5 above)
Comment (Devyani606) [reply app/workers/assessment_webhook_worker.rb]: done

File: app/controllers/api/v1/results_controller.rb [hunk 6]
```diff
@@ -28,25 +28,63 @@ def find_drive_candidates
...
-          @drives_candidates.each do |drives_candidate|
-            candidate = drives_candidate.candidate
-            submission = drives_candidate.submissions.order('total_marks desc').first
-            next unless submission
+          csv << ['First Name', 'Last Name', 'Email', 'Score'] + test_case_headers
+          chk_for_candidates_csv_results(csv, problem)
+        end
+      end
+
+      def chk_for_candidates_csv_results(csv, problem) #rubocop:disable all
```
Comment (anilmaurya) [inline app/controllers/api/v1/results_controller.rb:43]: We should refactor this method, it has duplicate code.

File: spec/controllers/api/v1/results_controller_spec.rb [hunk 7]
```diff
@@ -70,5 +72,15 @@
...
+    it 'returns single candidate result data in csv' do
+      get :csv_result, params: { drife_id: @drive.id, problem_id: @problem.id, drives_candidate_id: @drives_candidate.id }, format: :csv
+      actual_row = [['First Name', 'Kiran'], ['Last Name', 'Patil'], ['Email', 'kiran@gmail.com'],
+                    ['Score', nil], ['Test case 1 actual output', nil], ['Test case 1 expected output', 'hello']]
+      table = CSV.parse(File.read('result_file.csv'), headers: true)
+      table.by_row[0]
+      expect([['Email', 'kiran@gmail.com'], ['First Name', 'Kiran'], ['Last Name', 'Patil'], ['Score', nil], ['Test case 1 actual output', nil],
+              ['Test case 1 expected output', 'hello']]).to match_array(actual_row)
+    end
   end
```
Comment (anilmaurya) [inline spec/controllers/api/v1/results_controller_spec.rb:85]: Add test case for following scenarios:
1. Multiple candidates in one drive with one problem
  a) All candidates has done submission
  b) Some candidates has done submission

2. Multiple candidates in one drive with multiple problem
 a) Everyone has done submission for all problem
 b) Some candidates has done partial submission
 c) All have done partial submission

File: config/application.yml [hunk 8]
```diff
@@ -9,3 +9,5 @@ AWS_ACCESS_KEY_ID: 'SDJFKSADSAD'
 AWS_SECRET_ACCESS_KEY: 'AaC+bADJJRyKYf7uV893WIUSADJH'
 AWS_REGION: 'ap-south-1'
 S3_BUCKET_URL: 'https://oncot.s3.ap-south-1.amazonaws.com'
+PARAMS_AI_KEY: '8723AOSID234'
+MAIN_URL: 'https://oncot-api.joshsoftware.com'
```
Comment (anilmaurya) [inline config/application.yml:13]: PARAMS_AI_URL is missing

File: app/workers/assessment_webhook_worker.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    drive_candidate = DrivesCandidate.where(id: drive_candidate_id).first
+    drive = drive_candidate.drive
+    return unless drive_candidate.present?
+
+    params = { key: ENV['PARAMS_AI_KEY'], assessment_schedule_id: drive_candidate.uuid, score: drive_candidate.score,
+               status: 'completed', report: "#{ENV['MAIN_URL']}/api/v1/drives/#{drive.id}/results/csv_result?drives_candidate_id=#{drive_candidate_id}" }
+    status = ParamAiApi.new(params).post
+    raise StandardError, 'status is not 200' unless status == 200
+  end
+end
```
Comment (anilmaurya) [inline app/workers/assessment_webhook_worker.rb:16]: Test case for worker is missing

File: config/application.yml (hunk 8 above)
Comment (Devyani606) [reply config/application.yml:13]: done

**Review by anilmaurya — APPROVED** [review]

**Pull request merged**